Add reserved IPs to ipcache in XDS

joestringer opened the pull request:
These IPs were not being properly propagated to XDS, because they sidestep the kvstore sync logic. Fix this in a fairly minimal way (Create a function from the existing code that can be called, then call it).

Fixes: #3087 
Fixes: 7448e41 ("endpoint: sync endpoint IP-SecID map to kvstore")

joestringer committed:
ipcache: Shift NPHDS logic to envoy

There was a bunch of Network Policy Hosts Discovery Service logic
littering the ipcache logic, shift it into envoy.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Push reserved IP->Identity mappings to XDS

Previously we were only handling the BPF case and missing these IPs in
XDS, so presumably we would not apply L3/L4 policies correctly for the
XDS (eg Envoy sidecar) case.

CC: Ian Vernon <ian@cilium.io>
Fixes: 7448e41aa047 ("endpoint: sync endpoint IP-SecID map to kvstore")
Signed-off-by: Joe Stringer <joe@covalent.io>

rlenglet commented on pkg/envoy/resources.go:
```diff
@@ -27,9 +34,63 @@ const (
+	if modType == ipcache.Upsert {
+		endpointIPs, _ := ipcache.IPIdentityCache.LookupByIdentity(ipIDPair.ID)
+
+		// Update XDS Cache as well with the new identity to IP mapping.
```
Remove "as well".

rlenglet commented on pkg/envoy/resources.go:
```diff
@@ -27,9 +34,63 @@ const (
+	cache.updateForOldIdentity(ipIDPair.ID)
+}
+
+// updateXDSCacheForOldIdentity updates the XDS cache entry for old identity
```
Fix the comment. The function name changed.

rlenglet commented on pkg/envoy/resources.go:
```diff
@@ -27,9 +34,63 @@ const (
+// updateXDSCacheForOldIdentity updates the XDS cache entry for old identity
+// with the list of IPs that is stored in the IPIdentityCache.
+func (cache *NPHDSCache) updateForOldIdentity(oldIdentity identity.NumericIdentity) {
+	endpointIPs, isIdentityInCache := ipcache.IPIdentityCache.LookupByIdentity(oldIdentity)
```
You shouldn't need to call `ipcache`. You have enough state in `xds.Cache` to determine that locally.

joestringer replied:
Well, not exactly. This code assumes that `ipcache.IPIdentityCache` has been updated with the new IP->ID mapping, but the `NPHDSCache` has not yet been updated (because this code is supposed to update it). So it goes back up to the ipcache to get the latest list of IPs for the identity and pushes this into NPHDS.

rlenglet replied:
That's not needed. Again, you have all the information you need here.
You'll probably need to refactor this function and `OnIPIdentityCacheChange` above.
If `modType` is `Delete`, you need to first get the list of IPs associated with the ID from the cache, remove the IP from that list, and then: 1) if the new set is empty, delete the entry from the cache, otherwise 2) update the entry in the cache with the new list.

joestringer replied:
I see the point, the interface is broken if the callback from that layer has to go back up to that layer to fetch the relevant information.

ianvernon reviewed (dismissed):
@rlenglet captured all needed changes with his review. Thanks for this; makes the code much cleaner!

joestringer committed:
envoy: Handle IP->ID deletes inside cache

Don't reach back up to the ipcache to handle deletes like this, rather
look through the cache and update the entries in place.

Signed-off-by: Joe Stringer <joe@covalent.io>

rlenglet commented on pkg/envoy/resources.go:
```diff
@@ -27,9 +37,92 @@ const (
+			logfields.IPAddr:       ipIDPair.IP,
+			logfields.Identity:     ipIDPair.ID,
+			logfields.Modification: modType,
+		}).Warningf("Can't update NPHDS cache")
```
`Warningf` -> `Warn`
update -> lookup

rlenglet commented on pkg/envoy/resources.go:
```diff
@@ -27,9 +37,92 @@ const (
+	} else {
+		npHost = msg.(*envoyAPI.NetworkPolicyHosts)
+	}
+	if modType == ipcache.Upsert {
```
I think this would be cleaner as:

```
switch modType {
case ipcache.Upsert:
	if msg == nil {
		npHost = ...
	} else {
		npHost = msg.(*envoyAPI.NetworkPolicyHosts)
	}
	...
	cache.Upsert(...)
case ipcache.Delete:
	if msg == nil {
		// Doesn't exist; already deleted.
		return
	} else {
		cache.handleIPDelete(...)
	}
}
```

rlenglet commented on pkg/envoy/resources.go:
```diff
@@ -14,7 +14,17 @@
+import (
+	"sort"
+
+	envoyAPI "github.com/cilium/cilium/pkg/envoy/cilium"
```
What is wrong with the `cilium` package name? :)

joestringer replied:
copy+paste :-) (Also, consistency with other source files)

rlenglet commented on pkg/envoy/resources.go:
```diff
@@ -27,9 +37,92 @@ const (
+			logfields.IPAddr:       peerIP,
+			logfields.Identity:     peerIdentity,
+			logfields.Modification: ipcache.Delete,
+		}).Warningf("Can't find IP in NPHDS cache")
```
`Warningf` -> `Warn`

rlenglet commented on pkg/envoy/resources.go:
```diff
@@ -27,9 +37,92 @@ const (
+
+	// If removing this host would result in empty list, delete it.
+	// Otherwise, update to a list that doesn't contain the target IP
+	if len(npHost.HostAddresses) <= 1 && targetIndex >= 0 {
```
Remove `&& targetIndex >= 0`, it's redundant.

rlenglet commented on pkg/envoy/xds/cache.go:
```diff
@@ -254,3 +254,14 @@ func (c *Cache) GetResources(ctx context.Context, typeURL string, lastVersion *u
+// Lookup finds the resource corresponding to the specified typeURL and resourceName,
+// if available, and returns it. Otherwise, returns nil. If an error occurs while
+// fetching the resource, also returns the error.
+func (c *Cache) Lookup(typeURL string, resourceName string) (proto.Message, error) {
```
Please add short unit tests for this method, for both the positive and negative cases.

joestringer replied:
Added a couple of basic tests.

rlenglet reviewed (approved):
This is great, thanks!
I have relatively minor comments. Nothing really blocking.

joestringer committed:
xds: Add tests for cache.Lookup

Signed-off-by: Joe Stringer <joe@covalent.io>

raybejjani commented on pkg/envoy/resources.go:
```diff
@@ -27,9 +37,92 @@ const (
+// Service.
+type NPHDSCache struct {
+	*xds.Cache
+}
```
Out of scope of this review, and only out of curiosity. Is there a reason we embedded the type rather than `type NPHDSCache *xds.Cache` This way NPHDSCache also gets some functions from `xds.Cache` I think, or perhaps we have some future plans to add more fields?

joestringer replied:
No reason, I thought that they are functionally equivalent.

raybejjani replied:
I had to check to figure it out. It is unintuitive :/ `type B *A` hides the public functions of A, `type B {*A}` doesn't. (https://play.golang.org/p/ri6K5HzLRy4)
It doesn't matter much but it would mean that a `NPHDSCache` implements all interfaces that `*xds.Cache` does. What might be more problematic is that you might not be able to define a function on `NPHDSCache` that has the same name as `*xds.Cache` because go doesn't have method overloading. :/ today-I-remembered go things

rlenglet replied:
In this case we need this type to also be of type `*xds.Cache`. That’s intentional.

raybejjani reviewed (approved):
much +1
